File: compiler/Gocode/src/vericomp/hash/nullhash/nullhash.go
package null

import (
	. "vericomp"
	. "vericomp/hash"
)

const _NUM_ELTS = 2

type nullHash struct {
	cb *ConstraintBuilder
}

func NewNullHash(cb *ConstraintBuilder) *nullHash {
	n := new(nullHash)
	n.cb = cb
	return n
}

func (n *nullHash) GetNumHashBits() int {
	return _NUM_ELTS*HASH_FIELD_BITS
}

func (n *nullHash) CreateHash(createVar func(index int) string) *HashType {
	return NewHashType(createVar, _NUM_ELTS)
}

func (n *nullHash) HashValue(val string) []string {
	return n.hashBits(n.cb.Split(val, HASH_FIELD_BITS), false)
}

func (n *nullHash) HashValueBits(valBits []string) []string {
	return n.hashBits(valBits, true)
}

func (n *nullHash) hashBits(valBits []string, rev bool) []string {
	hashBits := make([]string, n.GetNumHashBits())
	start := 0
	if len(valBits) < len(hashBits) {
		start = len(hashBits) - len(valBits)
	}

	if rev {
		valBits = Rev(valBits)
	}
	
	copy(hashBits[start:], valBits)
	for i := 0; i < start; i++ {
		hashBits[i] = n.cb.Zero
	}
	
	return hashBits
}

func (n *nullHash) HashChildren(left, right []string) []string {
	return n.cb.Xor(left, right)
}

func (n *nullHash) Final(hashBits []string) *HashType {
	return NewHashTypeFinal(hashBits, _NUM_ELTS, n.cb)
}

func (n *nullHash) GetNumElts() int {
	return _NUM_ELTS
}

File: pepper/storage/null_hash.cpp
#include <storage/null_hash.h>
#include <storage/hasher.h>

NullHash::NullHash() {
}

NullHash::~NullHash() {

}

int NullHash::getNumHashBits() {
  return HashType::FIELD_SIZE*2;
}

HashType* NullHash::createHash(const Bits& hashBits) {
  return new HashType(hashBits, 2);
}

Bits NullHash::hash(const Bits& v) {
  Bits hashBits(getNumHashBits());

  uint32_t start = 0;
  if (v.size() < hashBits.size()) {
    start = hashBits.size() - v.size();
  }

  for (uint32_t i = 0; i < hashBits.size(); i++) {
    bool bit = (i < start) ? false : v[i - start];
    hashBits[i] = bit;
  }

  return hashBits;
}

Bits NullHash::hash(const Bits& left, const Bits& right) {
  return left ^ right;
}
